mlir/lib/Dialect/MQTOpt/Transforms/ElidePermutationsPattern.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2023 - 2025 Chair for Design Automation, TUM
+ * Copyright (c) 2025 Munich Quantum Software Company GmbH
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * Licensed under the MIT License
+ */
+
+#include "mlir/Dialect/MQTOpt/IR/MQTOptDialect.h"
+#include "mlir/Dialect/MQTOpt/Transforms/Passes.h"
+
+#include <llvm/ADT/STLExtras.h>
+#include <mlir/IR/MLIRContext.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/IR/Value.h>
+#include <mlir/IR/ValueRange.h>
+#include <mlir/Support/LLVM.h>
+#include <mlir/Support/LogicalResult.h>
+#include <vector>
+
+namespace mqt::ir::opt {
+
+/**
+ * @brief This pattern attempts to remove SWAP gates by re-ordering qubits.
+ */
+struct ElidePermutationsPattern final : mlir::OpRewritePattern<SWAPOp> {
+
+  explicit ElidePermutationsPattern(mlir::MLIRContext* context)
+      : OpRewritePattern(context) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(SWAPOp op, mlir::PatternRewriter& rewriter) const override {
+    const bool isControlledSwap = !op.getAllCtrlInQubits().empty();
+    if (isControlledSwap) {
+      return mlir::failure();
+    }
+
+    auto inQubits = op.getInQubits();
+
+    std::vector<mlir::Value> swappedQubits;
+    for (auto&& reversed : llvm::reverse(inQubits)) {
+      swappedQubits.push_back(reversed);
+    }
+
+    rewriter.replaceAllOpUsesWith(op, swappedQubits);
+
+    rewriter.eraseOp(op);
+
+    return mlir::success();
+  }
+};
+
+/**
+ * @brief Populates the given pattern set with the `ElidePermutationsPattern`.
+ *
+ * @param patterns The pattern set to populate.
+ */
+void populateElidePermutationsPatterns(mlir::RewritePatternSet& patterns) {
+  patterns.add<ElidePermutationsPattern>(patterns.getContext());
+}
```

Comment:
I know that this is only three lines of code, but I am still wondering if that boilerplate is necessary or if one could just include this as part of the tablegen definition somehow. Also totally fine if the answer to all of that is simply: no, it's necessary.
I just have the feeling that some of these patterns and passes are so simple that they could probably be written in tablegen itself (which has support for defining passes in some form).

Reply:
Since the `GateElimination` in #1140 has two patterns which need to be added, I don't think that it is possible (or worth it) to make this even shorter. But I'll have a look at it (together with the `ElidePermutations` class).